### PR TITLE
db: Deprecate the repo.cloned column

### PR DESCRIPTION
**Pull request by ryanslade**
We will be relying on gitserver_repos.clone_status instead.

This change moves us in that direction by:

1. Removing the cloned field from type.Repo
2. Falling back to read from repo.cloned ONLY when there is no
   corresponding row in gitserver_repos.

In the next release we will delete the repos.cloned column.

### CODE REVIEW

File: migrations/frontend/1528395798_add_clone_status_conditional_indexes.down.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+BEGIN;
+
+DROP INDEX IF EXISTS gitserver_repos_cloned_status_idx;
```
Comment (arussellsaw) [inline migrations/frontend/1528395798_add_clone_status_conditional_indexes.down.sql:3]: is dropping these indexes safe or necessary in this PR? might it be better to drop them in the PR that also removes the field?

File: internal/database/repos_db_test.go [hunk 2]
```diff
@@ -59,18 +58,30 @@ func createRepo(ctx context.Context, t *testing.T, db *sql.DB, repo *types.Repo)
 	}
 }
 
-func mustCreate(ctx context.Context, t *testing.T, db *sql.DB, repos ...*types.Repo) []*types.Repo {
+func mustCreate(ctx context.Context, t *testing.T, db *sql.DB, repo *types.Repo, cloned bool) []*types.Repo {
```
Comment (tsenart) [inline internal/database/repos_db_test.go:61]: Nit: Boolean parameters make it quite hard to read at a glance. It'd make it nicer if we'd pass in a `types.CloneStatus`

File: migrations/frontend/1528395798_add_clone_status_conditional_indexes.up.sql [hunk 3]
```diff
@@ -0,0 +1,8 @@
+BEGIN;
+
+CREATE INDEX IF NOT EXISTS gitserver_repos_cloned_status_idx ON gitserver_repos (repo_id) WHERE clone_status = 'cloned';
```
Comment (tsenart) [inline migrations/frontend/1528395798_add_clone_status_conditional_indexes.up.sql:3]: Can you leave a comment for posterity explaining why three separate partial indexes are better than a single one?

**Review by tsenart — APPROVED** [review]

File: migrations/frontend/1528395798_add_clone_status_conditional_indexes.down.sql (hunk 1 above)
Comment (arussellsaw) [reply migrations/frontend/1528395798_add_clone_status_conditional_indexes.down.sql:3]: i just realised this is the DOWN side of the migration 🤦

**Review by unknwon — APPROVED** [review]
LGTM, agree with @tsenart's comment about separate partial indexes.

File: internal/database/repos_db_test.go (hunk 2 above)
Comment (ryanslade) [reply internal/database/repos_db_test.go]: Good idea, done.

**Comment by daxmc99** [comment]
Are we still planning to deprecate this field? Running down a repo syncing issue and I struggled to determine if this is still used.

**Comment by ryanslade** [comment]
Yes, the field isn't used anymore. We just need to do the final cleanup that actually removes the column from the database

**Pull request merged**